### PR TITLE
Add granite

### PR DESCRIPTION
**Pull request by Minestar6**
# Add granite3.2
添加了granite3.2模型的PaddlePaddle 实现，包含模型代码、配置文件、单元测试和 下载支持。
# 对齐验证
## 精度对齐
前 20 个 token 的 logits 对齐结果如下：
* mean diff: 0.000000e+00
* max diff:  0.000000e+00

### 训练数据
* GSM8K SFT 数据
* 训练步数：300 steps
### 训练结果
image

其中ms-swift不支持granite模型，添加的主要注册代码如下：
```python
from transformers import AutoConfig, GraniteForCausalLM
from swift.llm.utils.model import LoRATM, MODEL_MAPPING, ModelType, get_model_tokenizer_with_flash_attn, register_model
from swift.llm.utils.template import TEMPLATE_MAPPING, Template, TemplateType, register_template

if not hasattr(ModelType, "granite"):
    ModelType.granite = "granite"
if not hasattr(TemplateType, "granite"):
    TemplateType.granite = "granite"



def get_model_tokenizer_granite(model_dir, torch_dtype, model_kwargs, load_model=True, **kwargs):
    model_config = AutoConfig.from_pretrained(model_dir, trust_remote_code=True)
    if hasattr(model_config, "pretraining_tp"):
        model_config.pretraining_tp = 1
    return get_model_tokenizer_with_flash_attn(
        model_dir,
        torch_dtype,
        model_kwargs,
        load_model,
        model_config=model_config,
        automodel_class=GraniteForCausalLM,
        **kwargs,
    )
    
if TemplateType.granite not in TEMPLATE_MAPPING:
    register_template(
        TemplateType.granite,
        Template(
            [],
            [
                "<|start_of_role|>user<|end_of_role|>{{QUERY}}<|end_of_text|>\n"
                "<|start_of_role|>assistant<|end_of_role|>"
            ],
            ["<|end_of_text|>\n"],
            ["<|end_of_text|>\n"],
            None,
            ["<|start_of_role|>system<|end_of_role|>{{SYSTEM}}<|end_of_text|>\n"],
        ),
        exist_ok=True,
    )

if ModelType.granite not in MODEL_MAPPING:
    register_model(
        ModelType.granite,
        "ibm-granite/granite-3.2-2b-instruct",
        LoRATM.llama,
        TemplateType.granite,
        get_model_tokenizer_granite,
        hf_model_id="ibm-granite/granite-3.2-2b-instruct",
        support_flash_attn=True,
        requires=["transformers>=4.45"],
        exist_ok=True,
    )
```

### CODE REVIEW

**Comment by a31413510** [comment]
验收通过，以下是代码review意见：

发现的问题（问题1需修复）

  🔴 问题1：GraniteForCausalLMPipe 缺少 _gen_aoa_config 委托

  文件: paddleformers/transformers/granite/modeling.py:589-591

  class GraniteForCausalLMPipe(GeneralModelForCausalLMPipe):
      config_class = GraniteConfig
      _decoder_layer_cls = GraniteDecoderLayer
      # 缺少以下关键属性！

  GraniteForCausalLMPipe 继承自 GeneralModelForCausalLMPipe 而非 GraniteForCausalLM，因此不会自动继承 _gen_aoa_config、_gen_inv_aoa_config、transpose_weight_keys 和
  _tied_weights_keys。当使用 Pipe 模式进行 flex_checkpoint 加载时，model_utils.py:2903 处会因 hasattr(cls, '_gen_aoa_config') == False 而报错。

  对比 OLMo2 的正确写法：

  class GraniteForCausalLMPipe(GeneralModelForCausalLMPipe):
      config_class = GraniteConfig
      _decoder_layer_cls = GraniteDecoderLayer
      _get_tensor_parallel_mappings = GraniteModel._get_tensor_parallel_mappings
      _init_weights = GraniteModel._init_weights
      _keep_in_fp32_modules = GraniteModel._keep_in_fp32_modules
      _tied_weights_keys = ["lm_head.weight"]
      transpose_weight_keys = GraniteModel.transpose_weight_keys
      _gen_aoa_config = GraniteForCausalLM._gen_aoa_config
      _gen_inv_aoa_config = GraniteForCausalLM._gen_inv_aoa_config

  ---
  🟡 问题2：GraniteDecoderLayer.forward 冗余逻辑（与 OLMo2 PR 相同问题）

  文件: paddleformers/transformers/granite/modeling.py:323-325

  outputs = (hidden_states,)
  if len(outputs) == 1 and isinstance(outputs, tuple):  # 永远为 True
      outputs = outputs[0]
  return outputs

  等价于 return hidden_states，建议直接简化。

  ---
  🟡 问题3：repeat_kv 本地重复定义

  文件: paddleformers/transformers/granite/modeling.py:61-66

  repeat_kv 在 paddleformers/nn/attention/utils.py 中已有实现，granite 模块又本地重复定义了一次。尽管它在 _flashmask_attention_forward
  中有实际使用，建议从公共模块导入而非重复定义。

  ---
  🟡 问题4：GraniteCompatibilityTest 中 setUpClass 缺少 torch.manual_seed
  文件: tests/transformers/granite/test_modeling.py:326-346
 setUpClass 创建模型时未设置随机种子（torch.manual_seed(0)），导致每次运行时随机权重不同，兼容性测试的数值对比可能非确定性。

建议：修复问题1（GraniteForCausalLMPipe 补全委托属性）后再合入，问题2-4 为 minor 级别建议。

**Review by nepeplwu — COMMENTED** [review]
建议更新下模型列表和矩阵能力：

https://github.com/PaddlePaddle/PaddleFormers/tree/develop#%E6%A8%A1%E5%9E%8B%E5%88%97%E8%A1%A8
https://github.com/PaddlePaddle/PaddleFormers/blob/develop/docs/zh/model_capability.md

**Comment by a31413510** [comment]
请处理代码冲突问题，重跑ci
image

File: paddleformers/transformers/granite/modeling.py [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+    _decoder_layer_cls = GraniteDecoderLayer
+    config_class = GraniteConfig
+    _decoder_layer_cls = GraniteDecoderLayer
+    _get_tensor_parallel_mappings = GraniteModel._get_tensor_parallel_mappings
+    _init_weights = GraniteModel._init_weights
+    _keep_in_fp32_modules = GraniteModel._keep_in_fp32_modules
+    _tied_weights_keys = ["lm_head.weight"]
+    transpose_weight_keys = GraniteModel.transpose_weight_keys
+    _gen_aoa_config = GraniteForCausalLM._gen_aoa_config
+    _gen_inv_aoa_config = GraniteForCausalLM._gen_inv_aoa_config
```
Comment (risemeup1111) [inline paddleformers/transformers/granite/modeling.py:671]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里的 Pipe 版本仍然走 `GeneralModelForCausalLMPipe` 的默认 `EmbeddingPipe/LMHeadPipe`，但普通路径里 `GraniteModel.forward()` 会显式做 `inputs_embeds *= embedding_multiplier`，`GraniteForCausalLM.forward()` 也会做 `logits /= logits_scaling`。现在 `GraniteForCausalLMPipe` 没有覆盖这两个阶段，所以 PP 路径会直接跳过这两处缩放，和非 Pipe 路径数值不一致。GraniteConfig 已经暴露了这些缩放参数，因此这不是纯粹的实现细节，而是会影响 PP 训练/推理一致性的行为差异。

建议补一个 Granite 专用的 embedding/head pipe，并把 `_rotary_emb_cls` 也显式接上，保证 Pipe 和非 Pipe 路径使用同一套缩放与 RoPE：

```python
class GraniteEmbeddingPipe(EmbeddingPipe):
    def forward(self, args):
        outputs = super().forward(args)
        if isinstance(outputs, tuple):
            return (outputs[0] * self.config.embedding_multiplier,) + outputs[1:]
        return outputs * self.config.embedding_multiplier

class GraniteLMHeadPipe(LMHeadPipe):
    def forward(self, args):
        logits = super().forward(args)
        return logits / self.config.logits_scaling

class GraniteForCausalLMPipe(GeneralModelForCausalLMPipe):
    _rotary_emb_cls = GraniteRotaryEmbedding
    _embedding_pipe_cls = GraniteEmbeddingPipe
    _lmhead_pipe_cls = GraniteLMHeadPipe
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查，发现一个需要先修复的 Pipe 路径一致性问题，细节在行级评论中。修复后我会再看一轮。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/granite/modeling.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/granite/modeling.py:671]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

我复查后确认，这次补丁只把 `logits_scaling` 接回来了，但 `embedding_multiplier` 仍然会在 tied-weights 分支里被绕开。`GeneralModelForCausalLMPipe` 在 `tie_word_embeddings=True` 时固定使用基类 `EmbeddingPipe`，不会走这里新增的 `GraniteEmbeddingPipe`；而 IBM 公开的 Granite 3.2 配置里 `tie_word_embeddings=true`，`embedding_multiplier=12.0`，所以真实 Granite 检查点的 PP 路径还是会和非 Pipe 路径不一致。

建议把 shared embedding 的装配也切到 `EmbeddingPipeCls`，或者在 Granite pipe 里单独覆盖这段装配逻辑：

```python
if config.tie_word_embeddings:
    self.add_sequential_layer(
        SharedLayerDesc(
            "model_shared_weight",
            EmbeddingPipeCls,
            shared_weight_attr="embedding_weight",
            config=config,
            embed_cls=self._embed_cls,
            rotary_emb_cls=self._rotary_emb_cls,
        ),
        "model",
    )
```

参考配置：<https://huggingface.co/ibm-granite/granite-3.2-8b-instruct/blob/main/config.json>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交，`logits_scaling` 这部分已经补上，但 Granite 在 tied-weights 配置下仍会绕过自定义 embedding pipe，导致 PP 路径和非 Pipe 路径不一致。详细说明已更新在现有行级讨论里，先把这个分支补齐再推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交，之前指出的 Granite PP 路径缩放一致性问题已经修复；没有发现新的需要阻塞合入的问题。CI 当前通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by zjjlivein — APPROVED** [review]
